Blocks/MMQ/mmpq.h

```diff
@@ -0,0 +1,171 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2016 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef BLOCKS_MMQ_MMPQ_H
+#define BLOCKS_MMQ_MMPQ_H
+
+// MMPQ is an in-memory priority queue, with the external interface loosely resembling the one of the original MMQ.
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+#include <set>
+
+#include "../SS/ss.h"
+
+#include "../../Bricks/time/chrono.h"
+
+namespace current {
+namespace mmq {
+
+template <typename MESSAGE, typename CONSUMER, size_t DEFAULT_BUFFER_SIZE = 1024, bool DROP_ON_OVERFLOW = false>
+class MMPQ {
+  static_assert(current::ss::IsEntrySubscriber<CONSUMER, MESSAGE>::value, "");
+
+ public:
+  // The type of messages to store and dispatch.
+  using message_t = MESSAGE;
+
+  // Consumer's `operator()` will be called from a dedicated thread, which is spawned and owned
+  // by the instance of MMPQ. See "Blocks/SS/ss.h" and its test for possible callee signatures.
+  using consumer_t = CONSUMER;
+
+  MMPQ(consumer_t& consumer) : consumer_(consumer), consumer_thread_(&MMPQ::ConsumerThread, this) {
+    consumer_thread_created_ = true;
+  }
+
+  // The destructor waits for the consumer thread to terminate, which implies committing all the queued messages.
+  ~MMPQ() {
+    if (consumer_thread_created_) {
+      CURRENT_ASSERT(consumer_thread_.joinable());
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        destructing_ = true;
+        condition_variable_.notify_all();
+      }
+      consumer_thread_.join();
+    }
+  }
+
+  // Adds a message to the buffer. Supports both copy and move semantics. THREAD SAFE.
+  using MutexLockStatus = current::locks::MutexLockStatus;
```

Comment:
I'm not against `using`, it's perfectly fine. I was speaking about flipping the lines to make the comment belong to the function it describes.